shared/fs/header.js

```diff
@@ -0,0 +1,167 @@
+import {Avatar, Box, ClickableBox, Icon, Text} from '../common-adapters'
+import {FolderHeaderBreadcrumbConnector} from './header-container'
+
+type FolderHeaderProps = {
+  path: Types.Path,
+}
+
+const stylesCommonRow = {
+  ...globalStyles.flexBoxRow,
+  alignItems: 'center',
+  borderBottomWidth: 1,
+  borderBottomColor: globalColors.black_05,
+  minHeight: isMobile ? 64 : 48,
+}
+
+const folderHeaderStyleRoot = {
+  ...stylesCommonRow,
+  alignItems: 'center',
+  justifyContent: 'center',
+}
+
+const folderHeaderStyleTree = {
+  ...stylesCommonRow,
+  alignItems: 'center',
+  paddingLeft: 16,
+  paddingRight: 16,
+}
+
+const folderBreadcrumbStyle = {
+  ...globalStyles.flexBoxRow,
+  alignItems: 'center',
+  paddingLeft: 0,
+  paddingRight: 0,
+}
+
+const rootHeaderContent = (
+  <Box style={folderHeaderStyleRoot}>
+    <Text type="BodyBig">Keybase Files</Text>
+  </Box>
+)
+
+type FolderHeaderState = {
+  pathElems: Array<string>,
+}
+
+type FolderBreadcrumbProps = {
+  index: number,
+  totalCrumbs: number,
+  visibility: Types.Visibility,
+  item: string,
+  path: Types.Path,
+}
+
+type FolderBreadcrumbConnectedProps = {
+  visibility: Types.Visibility,
+  item: string,
+  onOpen: () => void,
+}
+
+const headerPathStyleParent = {color: globalColors.black_60}
+const iconStyle = {
+  fontSize: 11,
+  marginLeft: globalMargins.xtiny,
+  marginRight: globalMargins.xtiny,
+}
+
+const styleTeamAvatar = {
+  marginRight: globalMargins.xtiny,
+}
+
+const FolderHeaderBreadcrumb = FolderHeaderBreadcrumbConnector(
+  ({visibility, item, onOpen}: FolderBreadcrumbConnectedProps) => {
+    return (
+      <ClickableBox onClick={onOpen}>
+        <Text type="BodySmallSemibold" style={headerPathStyleParent}>
+          {item}
+        </Text>
+      </ClickableBox>
+    )
+  }
+)
+
+const FolderHeaderItem = ({index, totalCrumbs, visibility, item, path}: FolderBreadcrumbProps) => {
+  let separator = <Icon type="iconfont-back" style={iconStyle} />
+  if (index === 0) {
+    separator = null
+  }
+  if (index === totalCrumbs - 1) {
+    return (
+      <Box style={folderBreadcrumbStyle}>
+        {separator}
+        {visibility === 'team' &&
+          index === 2 && <Avatar size={12} teamname={item} isTeam={true} style={styleTeamAvatar} />}
+        <Text type="BodyBig">{item}</Text>
+      </Box>
+    )
+  } else {
+    return (
+      <Box style={folderBreadcrumbStyle}>
+        {separator}
+        {visibility === 'team' &&
+          index === 2 && <Avatar size={12} teamname={item} isTeam={true} style={styleTeamAvatar} />}
+        <FolderHeaderBreadcrumb visibility={visibility} item={item} path={path} />
+      </Box>
+    )
+  }
+}
+
+class FolderHeader extends React.Component<FolderHeaderProps, FolderHeaderState> {
+  state: FolderHeaderState
+
+  constructor(props: FolderHeaderProps) {
+    super(props)
+    this.state = {
+      pathElems: Types.getPathElements(props.path),
```

Comment:
using state to hold derived props is mostly an anti-pattern. just pass in the values this thing needs as props. make this dumber and your container smarter